File: code/graph_algorithms/src/left_view_binary_tree/left_view_binary_tree.java
// Part of Cosmos by OpenGenus Foundation
// This class prints the left nodes of a binary tree
class Node {
    int  data;
    Node left;
    Node right;  
    
    public Node(int node) {
        data = node;
        left = right = null;
    }
}

class BinaryTree {
    Node root; // Define root node
    int max_level = 0; // Define max level
 
    // Recursive function that prints left nodes of tree
    void printLeftView(Node node, int cur_level) {
        if (node == null) return;      

        if (max_level < cur_level) {
            System.out.print(node.data + " ");
            max_level = cur_level; 
        }
 
        // Recurisely traverse along the left and right nodes of child node
        printLeftView(node.left, cur_level + 1);
        printLeftView(node.right, cur_level + 1);
    }

    void leftView() {
        printLeftView(root, 1);
    }
 
    public static void main(String args[]) {
        // Binary tree test
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);
        tree.root.right.left = new Node(6);
        tree.root.right.right = new Node(7);
        tree.root.left.left.left = new Node(8);
        System.out.print("Left nodes of binary tree: ");
        tree.leftView();     
    }
}

File: code/graph_algorithms/src/left_view_binary_tree/left_view_binary_tree.cpp
#include <iostream>
using namespace std;
// Part of Cosmos by OpenGenus Foundation
struct node
{
    int d;
    node *left;
    node *right;
};

struct node* newnode(int num)
{
    node *temp = new node;
    temp->d = num;
    temp->left = temp->right = NULL;
    return temp;
}

void lefttree(struct node *root, int level, int *maxlevel)
{
    if (root == NULL)
        return;


    if (*maxlevel < level)
    {
        cout << root->d << endl;
        *maxlevel = level;
    }

    lefttree(root->left, level + 1, maxlevel);
    lefttree(root->right, level + 1, maxlevel);

}

int main()
{
    node *root = newnode(12);
    root->left = newnode(10);
    root->right = newnode(30);
    root->right->left = newnode(25);
    root->right->right = newnode(40);

    int maxlevel = 0;

    lefttree(root, 1, &maxlevel);

    cout << endl;
    return 0;
}
